### ISSUE TITLE
how to deal with "too many open files"

### ISSUE THREAD

**Issue by martinlippert** [oracle/graal#1581]
I am experimenting with compiling an app down to a native image. This app opens a lot of files, so I am running into "too many open files" exceptions. Is there a way to avoid that and to allow the native image to behave in a similar way than the regular JVM, where I can easily open many many files?

**Comment by bobmcwhirter** [comment]
Are you simply exceeding some OS `ulimit` or do you imagine there's a file descriptor leak within Substrate itself, or your app?

Can you post some stacktraces if the "too many open files" error seems inexplicable?

**Comment by martinlippert** [comment]
I think this is directly related to the `ulimit` (there doesn't seem to be a leak in Substrate nor my app), but the regular JVM doesn't seem to be affected, so I am trying to understand the difference... :-)

**Comment by martinlippert** [comment]
For completeness, the first error that I see is:

```
java.io.FileNotFoundException: <...myfilename...>
	at com.oracle.svm.core.posix.PosixUtils.fileOpen(PosixUtils.java:235)
	at java.io.FileOutputStream.open(FileOutputStream.java:626)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at java.io.FileWriter.<init>(FileWriter.java:90)
```

Shortly after that, I see a

```
java.nio.file.FileSystemException: <...somefolder...>: Too many open files
	at sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:407)
	at java.nio.file.Files.newDirectoryStream(Files.java:457)
	at java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:300)
	at java.nio.file.FileTreeWalker.walk(FileTreeWalker.java:322)
	at java.nio.file.FileTreeIterator.<init>(FileTreeIterator.java:72)
	at java.nio.file.Files.walk(Files.java:3574)
	at java.nio.file.Files.walk(Files.java:3625)
```

And running `lsof -p <pid>` reveals that the process has indeed about 256 file handles in use, which is the `ulimit` setting here.

I could try to find out how to increase that setting on my system, but I would very much prefer to find an alternative solution. Would love to run the native image in different environments and can't assume specific ulimit settings there.

**Comment by bobmcwhirter** [comment]
ulimit is applied by the OS so we can’t subvert that. I’m reviewing to see if we aren’t leaking something though. Seems wrong that a way would have that many files open concurrently.

**Comment by martinlippert** [comment]
@bobmcwhirter I think the app that I am dealing with does indeed keep many files open (partly because of its nature), so don't worry so much about Substrate leaking something. I don't think that is the case here. The same number of files are opened when I run the app on the regular JVM. Interestingly it looks like a different ulimit is applied to that somehow.

**Comment by peter-hofer** [comment]
HotSpot (the Oracle/OpenJDK VM) seems to always increase the file descriptor limit from the current soft limit to the absolute hard limit, which is why it does not fail here. 256 seems rather low however, on my machine the soft limit is 1024. What's the hard limit on your system (`ulimit -aH`, 4096 here)?

**Comment by martinlippert** [comment]
`open files                      (-n) unlimited`

Should maybe the native-image do the same than HotSpot with regards to increasing the file descriptor limit?

**Comment by peter-hofer** [comment]
Yes, that would avoid surprises like these where the code runs fine on HotSpot, but where a native image hits the open file limit. I just wonder to what extent a program is supposed to respect the environment's current soft limit. HotSpot is a more heavyweight runtime environment to start with, but I'm not sure if we should apply the same rules to native images.

**Comment by martinlippert** [comment]
Yeah, agree, maybe it is not perfect to apply this rule automatically to every native image. Maybe worth an option for the native-image builder?

**Comment by peter-hofer** [comment]
Fixed in 2cf6310e9dcb32ae3198489094c7ada655a77488: we now always raise the limit to the hard maximum.

**Comment by martinlippert** [comment]
Awesome, thanks!!!
